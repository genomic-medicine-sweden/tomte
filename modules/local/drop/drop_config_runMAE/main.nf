process DROP_CONFIG_RUN_MAE {
    tag "DROP_CONFIG_RUN_MAE"
    label 'process_drop'

    input:
    tuple val(meta), path(fasta), path(fai)
    tuple val(meta2), path(gtf)
    tuple val(meta), path(dict)
    path sample_annotation
    val(genome)
    tuple path(bam), path(bai)
    tuple path(vcf), path(vcf_tbi)
    tuple path(drop_mae_high_q_vcf), path(drop_mae_high_q_vcf_tbi)

    output:
    path('config.yaml')       , emit: config_drop
    path('output')            , emit: drop_mae_out
    path('MAE_results_*')     , emit: drop_mae_tsv
    path('gene_name_mapping*'), emit: drop_gene_name
    path "versions.yml"       , emit: versions

    when:
    // Exit if running this module with -profile conda / -profile mamba
    workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() == 0 ||
        { log.error("Local DROP module does not support Conda. Please use Docker / Singularity / Podman instead."); return false }

    script:
    def args = task.ext.args ?: ''
    def genome_assembly = "${genome}".contains("h37") ? "hg19" : "${genome}"
    def gtf_basename = gtf.getName().replaceAll(/\.(gtf|gff3?|GTF|GFF3?)$/, '')

    """
    TMPDIR=\$PWD
    HOME=\$PWD

# Add the gtf to GENE_ANNOTATION column (necessary for samples that will be run through MAE)
    awk -v val="${gtf_basename}" 'BEGIN{FS=OFS="\t"}
    NR==1 {for(i=1;i<=NF;i++) if(\$i=="GENE_ANNOTATION") col=i}
    NR>1 && \$col=="NA" {\$col=val}
    {print}
    ' ${sample_annotation} > tmp && mv tmp ${sample_annotation}

    drop init

    drop_config.py \\
        --genome_fasta ${fasta} \\
        --gtf ${gtf} \\
        --drop_module MAE \\
        --genome_assembly $genome_assembly \\
        --drop_mae_high_q_vcf $drop_mae_high_q_vcf \\
        --output config.yaml

    snakemake mae --cores ${task.cpus} --rerun-triggers mtime $args

    cp output/processed_results/mae/mae/MAE_results_*.tsv .
    rm MAE_results_*_rare.tsv
    cp output/processed_data/*/gene_name_mapping_*.tsv .

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        drop_config: \$(drop_config.py --version)
        drop: \$(echo \$(drop --version) |  sed -n 's/drop, version //p')
    END_VERSIONS
    """

    stub:
    """
    touch config.yaml
    touch MAE_results_genome.tsv
    touch gene_name_mapping_.tsv
    mkdir output

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        drop_config: \$(drop_config.py --version)
        drop: \$(echo \$(drop --version) |  sed -n 's/drop, version //p')
    END_VERSIONS
    """
}
