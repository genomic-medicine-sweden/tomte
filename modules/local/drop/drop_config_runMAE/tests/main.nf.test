nextflow_process {

    name "Test Process DROP_CONFIG_RUN_MAE"
    script "../main.nf"
    process "DROP_CONFIG_RUN_MAE"
    config "./nextflow.config"
    tag "modules"
    tag "modules_local"
    tag "drop"
    tag "drop_config_runMAE"

    setup {
        run("GUNZIP", alias: "GUNZIP_FASTA") {
            script "../../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'fasta' ],
                    file("${projectDir}/test_data/grch37_chr21.fa.gz", checkIfExists: true)
                ])
                """
            }
        }

        run("SAMTOOLS_FAIDX") {
            script "../../../../../modules/nf-core/samtools/faidx/main.nf"
            process {
                """
                input[0] = GUNZIP_FASTA.out.gunzip
                input[1] = [[],[]]
                input[2] = []
                """
            }
        }

        run("GUNZIP", alias: "GUNZIP_GTF") {
            script "../../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'gtf' ],
                    file("${projectDir}/test_data/grch37_chr21.gtf.gz", checkIfExists: true)
                ])
                """
            }
        }

        run("DROP_SAMPLE_ANNOT") {
            script "../../../../../modules/local/drop/drop_sample_annot/main.nf"
            process {
                """
                input[0] = Channel.of([
                            ['ACC5963A1'], [false], ['reverse'], ['NA'], ["${projectDir}/test_data/test_ACC5963A1.vcf.gz"], ["${projectDir}/test_data/test_ACC5963A1.vcf.gz.tbi"],  ['ACC5963A1'],
                            file("${projectDir}/test_data/ACC5963A1.bam", checkIfExists: true),
                            file("${projectDir}/test_data/ACC5963A1.bam.bai", checkIfExists: true)
                        ])
                input[1] = channel.fromPath("${projectDir}/test_data/drop_data/geneCounts.tsv.gz", checkIfExists: true)
                input[2] = channel.fromPath("${projectDir}/test_data/drop_data/sampleAnnotation.tsv", checkIfExists: true)
                input[3] = 'outrider'
                input[4] = 'fraser'
                """
            }
        }

        run("GATK4_CREATESEQUENCEDICTIONARY", alias: "BUILD_DICT") {
        script "../../../../../modules/nf-core/gatk4/createsequencedictionary/main.nf"
        process {
            """
            input[0] = GUNZIP_FASTA.out.gunzip
            """
            }
        }

    }

    test("Should run without failures") {

        when {
            process {
                """
                input[0] = GUNZIP_FASTA.out.gunzip.join(SAMTOOLS_FAIDX.out.fai).collect()
                input[1] = GUNZIP_GTF.out.gunzip.collect()
                input[2] = BUILD_DICT.out.dict.collect()
                input[3] = DROP_SAMPLE_ANNOT.out.drop_annot
                input[4] = 'GRCh37'
                input[5] = Channel.of([
                        file("${projectDir}/test_data/ACC5963A1.bam", checkIfExists: true),
                        file("${projectDir}/test_data/ACC5963A1.bam.bai", checkIfExists: true)
                    ])
                input[6] = Channel.of([
                        file("${projectDir}/test_data/test_ACC5963A1.vcf.gz", checkIfExists: true),
                        file("${projectDir}/test_data/test_ACC5963A1.vcf.gz.tbi", checkIfExists: true)
                    ])
                input[7] = Channel.of([
                        file("${projectDir}/test_data/qc_vcf_1000G_hg19_chr21.vcf.gz", checkIfExists: true),
                        file("${projectDir}/test_data/qc_vcf_1000G_hg19_chr21.vcf.gz.tbi", checkIfExists: true)
                    ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.config_drop.collect{ file(it).name },
                    process.out.drop_gene_name,
                    process.out.drop_mae_out.collect{ file(it).name },
                    process.out.drop_mae_tsv.collect{ file(it).name },
                    process.out.versions,
                ).match()}
            )
        }

    }

}
