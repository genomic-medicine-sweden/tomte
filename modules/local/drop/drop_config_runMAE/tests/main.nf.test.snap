{
    "Should run without failures": {
        "content": [
            [
                "config.yaml"
            ],
            [
                "gene_name_mapping_grch37_chr21.tsv:md5,f70b29e044d02ac3d2ae6e83221aa978"
            ],
            [
                "output"
            ],
            [
                "MAE_results_grch37_chr21.tsv"
            ],
            [
                "versions.yml:md5,eeac97ccc224807222bb8768daaff3c6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.5"
        },
        "timestamp": "2025-09-16T14:40:17.893217795"
    }
}